### PR TITLE
Feature/authorization

### PR DESCRIPTION
**Pull request by tsohleDev**
# Authorization

In this PR we:
- Install CanCanCan in your project.
- Add a `role` column to the users table. Remember to use a migration for this.
- A user can delete a post if it is theirs or if they have an admin role (column `role` has value `"admin"`). Use CanCanCan for this authorization.
    - For that you need to implement the post deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.
- A user can delete a comment if it is theirs or if they have an admin role (column `role` has value `"admin"`). Use CanCanCan for this authorization.
    - For that you need to implement the comment deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -1,14 +1,20 @@
 class PostsController < ApplicationController
+  before_action :authenticate_user!
+
```
Comment (ishimwezachee) [inline app/controllers/posts_controller.rb]: - [x]  You did great job to configure Cancan to implement the authorization in you project.  and I like how you are using `authorize!` method to check for the authority on destroy method in both posts and comment controller actions.  One thing I would suggest you to use `load_and_authorize_resource `method to handle the loading and authorize other actions in your post controller.
`load_and_authorize_resource `is a convenient way to automatically load and authorize resources for multiple actions in a controller

File: app/models/user.rb [hunk 2]
```diff
@@ -1,4 +1,8 @@
 class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable, :confirmable
   has_many :posts, foreign_key: :author_id, dependent: :destroy
   has_many :comments, foreign_key: :author_id, dependent: :destroy
   has_many :likes, foreign_key: :author_id, dependent: :destroy
```
Comment (ishimwezachee) [inline app/models/user.rb:8]: - [ OPTIONAL] For credits you can even create the `user.admin? `method inside the user model, it is not a  part of the requirement, but you never know, in the future you might like to add the admin features to your project.

**Review by ishimwezachee — CHANGES_REQUESTED** [review]
Hi @tsohleDev  and @MICHAELKITH , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
image

## Highlights
✅ CanCan is configured correctly 
✅ using authorize! method to check authorization within a specific action

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (tsohleDev) [reply app/controllers/posts_controller.rb]: Yes convenience... Thank @ishimwezachee

**Review by KDlamini — APPROVED** [review]
### Hi team @tsohleDev, @MICHAELKITH 🙋‍♀️,  

Excellent work! 💯 
You have done a great job implementing all the required changes. 🎊 🥳 
Your feature is complete! There is nothing else to say other than... it's time to merge it. :shipit: 
Congratulations!!!

## Highlights 🌞 

✔️  Good PR title and description
✔️  Users are only authorized to delete posts and comments they own
✔️  The `role` column exists in the user's table
✔️  There's a CanCanCan `ability` class that defines the set of rules for user-post interactions
✔️  Project has **CanCan** Gem in **Gemfile**

# Status: Approved ✔️ 

oscars-standing-ovation
 
Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @KDlamini  in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**